kernel/src/hil/display.rs

```diff
@@ -0,0 +1,376 @@
+            (Rotation::Rotated270, Rotation::Rotated180) => Rotation::Rotated90,
+            (Rotation::Rotated270, Rotation::Rotated270) => Rotation::Rotated180,
+        }
+    }
+}
+
+impl Sub for Rotation {
+    type Output = Self;
+
+    fn sub(self, other: Self) -> Self {
+        match (self, other) {
+            (_, Rotation::Normal) => self,
+
+            (Rotation::Normal, Rotation::Rotated90) => Rotation::Rotated270,
+            (Rotation::Normal, Rotation::Rotated180) => Rotation::Rotated180,
+            (Rotation::Normal, Rotation::Rotated270) => Rotation::Rotated90,
+
+            (Rotation::Rotated90, Rotation::Rotated90) => Rotation::Normal,
+            (Rotation::Rotated90, Rotation::Rotated180) => Rotation::Rotated270,
+            (Rotation::Rotated90, Rotation::Rotated270) => Rotation::Rotated180,
+
+            (Rotation::Rotated180, Rotation::Rotated90) => Rotation::Rotated90,
+            (Rotation::Rotated180, Rotation::Rotated180) => Rotation::Normal,
+            (Rotation::Rotated180, Rotation::Rotated270) => Rotation::Rotated270,
+
+            (Rotation::Rotated270, Rotation::Rotated90) => Rotation::Rotated180,
+            (Rotation::Rotated270, Rotation::Rotated180) => Rotation::Rotated90,
+            (Rotation::Rotated270, Rotation::Rotated270) => Rotation::Normal,
+        }
+    }
+}
+
+#[derive(Copy, Clone, PartialEq)]
+#[allow(non_camel_case_types)]
+#[non_exhaustive]
+pub enum PixelFormat {
+    /// Pixels encoded as 1-bit, used for monochromatic displays
+    Mono,
+    /// Pixels encoded as 2-bit red channel, 3-bit green channel, 3-bit blue channel.
+    RGB_233,
+    /// Pixels encoded as 5-bit red channel, 6-bit green channel, 5-bit blue channel.
+    RGB_565,
+    /// Pixels encoded as 8-bit red channel, 8-bit green channel, 8-bit blue channel.
+    RGB_888,
+    /// Pixels encoded as 8-bit alpha channel, 8-bit red channel, 8-bit green channel, 8-bit blue channel.
+    ARGB_8888,
+    // other pixel formats may be defined.
+}
+
+impl PixelFormat {
+    pub fn get_bits_per_pixel(&self) -> usize {
+        match self {
+            Self::Mono => 1,
+            Self::RGB_233 => 8,
+            Self::RGB_565 => 16,
+            Self::RGB_888 => 24,
+            Self::ARGB_8888 => 32,
+        }
+    }
+}
+
+#[derive(Copy, Clone, PartialEq)]
+#[allow(non_camel_case_types)]
+#[non_exhaustive]
+pub enum CharacterFormat {
+    /// Characters are encoded using 8 bits ASCII, used for monochromatic displays
+    ASCII,
+    /// Characters are encoded using UTF8, used for monochromatic displays
+    UTF8,
+    /// Characters are encoded using 16 bits, as in the VGA text mode
+    /// 0   1   2   3   4   5   6   7   8   9   10  11  12  13  14  15  16
+    /// +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+    /// | B |Background | Foreground    |          Code Point           |
+    /// +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+    /// B - blink
+    VGA,
```

Comment:
What encoding in VGA? What happens if the code point is multiple bytes? What endianness?